sklearn/datasets/rcv1.py

```diff
@@ -0,0 +1,213 @@
+    targets_path = join(rcv1_dir, "targets")
+    sample_id_path = join(rcv1_dir, "sample_id")
+    categories_path = join(rcv1_dir, "categories")
+    available = exists(samples_path)
+
+    if download_if_missing and not available:
+        _mkdirp(rcv1_dir)
+
+        # load data (X) and sample_id
+        file_urls = ["%s_test_pt%d.dat.gz" % (URL, i) for i in range(4)]
+        file_urls.append("%s_train.dat.gz" % URL)
+        files = []
+        for file_url in file_urls:
+            logger.warning("Downloading %s" % file_url)
+            f = BytesIO(urlopen(file_url).read())
+            files.append(GzipFile(fileobj=f))
+
+        Xy = load_svmlight_files(files, n_features=N_FEATURES)
+
+        # Training data is before testing data
+        X = sp.vstack([Xy[8], Xy[0], Xy[2], Xy[4], Xy[6]]).tocsr()
+        sample_id = np.hstack((Xy[9], Xy[1], Xy[3], Xy[5], Xy[7]))
+        sample_id = sample_id.astype(np.int32)
+
+        # load target (y) and sample_id_bis
+        logger.warning("Downloading %s" % URL_topics)
+        f = BytesIO(urlopen(URL_topics).read())
+        f = GzipFile(fileobj=f, mode='r')
+
+        n_cat = -1
+        n_doc = -1
+        doc_previous = -1
+        y = np.zeros((N_SAMPLES, N_CATEGORIES), dtype=np.int8)
+        sample_id_bis = np.zeros(N_SAMPLES, dtype=np.int32)
+        category_names = {}
+        for line in f.readlines():
+            if len(line.split(" ")) > 2:
+                cat, doc, _ = line.split(" ")
+                if cat not in category_names:
+                    n_cat += 1
+                    category_names[cat] = n_cat
+
+                doc = int(doc)
+                if doc != doc_previous:
+                    doc_previous = doc
+                    n_doc += 1
+                    sample_id_bis[n_doc] = doc
+                y[n_doc, category_names[cat]] = 1
+
+        # Samples in X are ordered with sample_id,
+        # whereas in y, they are ordered with sample_id_bis.
+        p = _find_permutation(sample_id_bis, sample_id)
+        y = sp.csr_matrix(y[p, :])
+
+        # save category names in a list, with same order than y
+        categories = [None] * N_CATEGORIES
+        for k in category_names.keys():
+            categories[category_names[k]] = k
+
+        joblib.dump(X, samples_path, compress=9)
+        joblib.dump(y, targets_path, compress=9)
+        joblib.dump(sample_id, sample_id_path, compress=9)
+        joblib.dump(categories, categories_path, compress=9)
+
+    try:
+        X, y, samples_id, categories
+    except NameError:
+        X = joblib.load(samples_path)
+        y = joblib.load(targets_path)
+        sample_id = joblib.load(sample_id_path)
+        categories = joblib.load(categories_path)
+
+    if shuffle:
+        X, y, sample_id = shuffle_(X, y, sample_id, random_state=random_state)
+
+    return Bunch(data=X, target=y, sample_id=sample_id,
+                 categories=categories, DESCR=__doc__)
+
+
```

Comment:
`_mkdirp` is duplication. I will change it.
The two permutation functions are from me.